src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -586,6 +586,68 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
     );
   };
 
+  createPendingBlock = async (previousBlock: Block) => {
+    const options = this.#options;
+    const minerOpts = options.miner;
+    // "instamine" mode means blockTime = 0
+    const instamine = this.#instamine;
+    let timestamp: number = null;
+    if (!instamine) {
+      // we can't always get the pending block's timestamp right because they
+      // could have the miner paused, but if blockTime > 0, blocks are mined at
+      // regular intervals, so we can at least make an educated guess
+      timestamp =
+        previousBlock.header.timestamp.toNumber() + minerOpts.blockTime;
+    }
+    const nextBlock = this.#readyNextBlock(previousBlock, timestamp);
+    // If the executables has inProgress transactions, we were already
+    // in the middle of mining a block when the pending block was requested.
+    // The block that is currently being mined is what we want to return as pending,
+    // so we'll reset the executables to be what it was when mining started,
+    // i.e. all inProgress executables are moved to the front of the pending executables.
+    const executables =
+      this.transactions.transactionPool.cloneAndResetExecutables();
+
+    // the number of transactions that should be included in this block depends
+    // on a few factors. in "instamine" mode (meaning blockTime = 0), we usually
+    // only mine 1 transaction per block. However, if the miner is currently
+    // paused (from miner_stop being called), whenever they restart the miner
+    // all executable transactions will be mined, so the pending block should
+    // contain all executables that will fit in a block. if we're not in
+    // instamine mode, the miner will always try to fill the block, regardless
+    // of whether the miner is paused or not
+    const paused = this.#isPaused();
+    const maxTransactions =
+      instamine && !paused ? Capacity.Single : Capacity.FillBlock;
+
+    // we don't want any events from mining to be
+    // caught by `this.vm`, so we need to make a new one.
+    const vm = await this.createVmFromStateTrie(
+      this.trie.copy(false),
+      options.chain.allowUnlimitedContractSize,
+      true
+    );
+
+    const miner = new Miner(minerOpts, executables, vm, this.#readyNextBlock);
+    let pendingBlock: Block;
+    // set up listener to actually assign the newly mined pending block
+    miner.on(
+      "block",
+      async (blockData: {
+        block: Block;
+        serialized: Buffer;
+        storageKeys: StorageKeys;
+        transactions: TypedTransaction[];
+      }) => {
+        pendingBlock = blockData.block;
+      }
+    );
+    // await so we're sure the above line has had time to set the pending block
+    // even though we don't actually need the result of `miner.mine`
+    await miner.mine(nextBlock, maxTransactions, true);
```

Comment:
I'm guessing that we can be confident that event handlers complete before resolving the return value of `miner.mine`, but would it be better to directly await the block?

Something like:
```javascript
const blockPromise = new Promise<Block>((resolve, reject) => {
miner.on(
      "block",
      async (blockData: {
        block: Block;
        serialized: Buffer;
        storageKeys: StorageKeys;
        transactions: TypedTransaction[];
      }) => {
        resolve(blockData.block);
      }
    );
});
miner.mine(nextBlock, maxTransactions, true);
return await blockPromise;
````

Reply:
Wayyyyy better, fixed!